runtime/stdlib.js

```diff
@@ -358,90 +358,203 @@ function caml_floatarray_create(len){
   return b
 }
 
+//Provides: caml_compare_val_tag
+//Requires: MlBytes
+function caml_compare_val_tag(a){
+  if (typeof a === "number") return 1000; // int_tag (we use it for all numbers)
+  else if (a instanceof MlBytes) return 252; // string_tag
+  else if (a instanceof Array && a[0] === (a[0]>>>0) && a[0] <= 255) {
+    // Look like an ocaml block
+    var tag = a[0] | 0;
+    // ignore double_array_tag because we cannot accurately set
+    // this tag when we create an array of float.
+    return (tag == 254)?0:tag
+  }
+  else if (a instanceof String) return 1252; // javascript string, like string_tag (252)
+  else if (typeof a == "string") return 1252; // javascript string, like string_tag (252)
+  else if (a instanceof Number) return 1000; // int_tag (we use it for all numbers)
+  else if (a.caml_custom) return 1255; // like custom_tag (255)
```

Comment:
@hhugo, this breaks when `a` is `null` or `undefined`

Reply:
https://github.com/ocsigen/js_of_ocaml/pull/920